service/common/src/main/java/org/apache/polaris/service/exception/IcebergExceptionMapper.java

```diff
@@ -165,12 +167,16 @@ public static Collection<String> getAccessDeniedHints() {
   }
 
   static int mapExceptionToResponseCode(RuntimeException rex) {
-    Optional<Throwable> cloudException =
-        Arrays.stream(ExceptionUtils.getThrowables(rex))
-            .filter(IcebergExceptionMapper::isCloudException)
-            .findAny();
-    if (cloudException.isPresent()) {
-      return mapCloudExceptionToResponseCode(cloudException.get());
+    for (Throwable t : ExceptionUtils.getThrowables(rex)) {
+      // Cloud exceptions can be wrapped by the Iceberg SDK
+      if (isCloudException(t)) {
+        return mapCloudExceptionToResponseCode(t);
+      }
+
+      // UnknownHostException isn't a RuntimeException so it's always wrapped
+      if (t instanceof UnknownHostException && t.getMessage().contains(AZURE_STORAGE_URL_SUFFIX)) {
+        return Status.NOT_FOUND.getStatusCode(); // Return a 404 to be consistent with S3/GCS
+      }
```

Comment:
Can we put this in `mapCloudExceptionToResponseCode`?

Reply:
`UnknownHostException` is not specific to any cloud SDK. I think it's fine to have it as a special case here.

Reply:
Currently that method is a bit specific to handling cloud exception types like S3Exception, AzureException, etc, which this is not. It's not super obvious to me that there's a clean way to refactor this while keeping existing behavior, in a way that we won't have to do more refactoring the future. This is just a very particular piece of code.

Reply:
It seems like the `doesAnyThrowableContainAccessDeniedHint` check is very similar, no? It has nothing to do with a cloud SDK. On the other hand, `AZURE_STORAGE_URL_SUFFIX` is indeed specific to a cloud.

Reply:
You're right but the current behavior is to only check the access denied hints for cloud exception types. Whether that's good or not I guess is its own discussion. There's some inherent difficulty here by playing a reactive role to exceptions and not knowing them all ahead of time.

Reply:
`AZURE_STORAGE_URL_SUFFIX` looks as cloud-specific as any exception type -- I think my reaction is mostly to having that relatively low-level detail in the top-level method `mapExceptionToResponseCode`

Reply:
This is definitely a cloud exception but above I meant the 3 top level cloud exception types as in S3Exception, AzureException, StorageException.

I did a bit of refactoring, lmk what you think.